designs/0032-cmdstan-output-formats.md

```diff
@@ -0,0 +1,230 @@
+and information structure.
+
+There are three sources of information:
+
+1. Stan program outputs: estimates for all model parameters, transformed parameters, and generated quantities.
+
+2. Inference engine outputs: various diagnostics.
+For the NUTS-HMC sampler, the current outputs are the stepsize and metric, and the per-iteration sampler variables (state).
+The optimization and variational algorithms also report on their respective iterations.
+
+3. Interface-level outputs:
+   + model name, compile options, Stan compiler and (Cmd)Stan version and compile options.
+   + inference algorithm, user-specified configuration options, and default config.
+   + input and output data descriptors.
+   + process-level information:  chain, iteration.
+   + timing information: timestamp information reported by processing events.
+
+There are two general kinds of information structure: tabular and hierarchical.
+[CSV files](https://en.wikipedia.org/wiki/Comma-separated_values) are tabular,
+as are R and Python dataframes and database tables.
+JSON is a [structured data formats](https://en.wikipedia.org/wiki/Data_exchange#Data_exchange_languages)
+which can represent hierarchical structures and arrays.
+
+## Current implementation
+
+We review the way in which the current Stan CSV file is generated for a run of the NUTS-HMC sampler.
+
+### Output mechanism:  callback writers
+
+Outputs are handled by objects of base class `stan::callbacks::writer`
+which is constructed from an output stream.
+Through function call operator overloading, the callback writer class
+methods provide the appropriate formatting for different kinds of data.
+When the user sends a request to the interface to do some kind of inference given a model and data,
+the interfaces first instantiate the Stan model object and the writer callbacks,
+then pass these in to `stan::services` functions which invoke the appropriate algorithm.
+The use of writer callbacks keeps the inference algorithms output-format agnostic.
+The current `stan::services` layer utilities are completely Stan CSV format-centric and will need to be refactored.
+
+
+### Stan program outputs
+
+The Stan model class `write_array` method returns a vector of doubles for all
+parameter, transformed parameter, and generated quantities (non-local) variables,
+in order of declaration in the Stan program.
+
+The model class methods `constrained_param_names`, `unconstrained_param_names`, `get_constrained_sizedtypes`, `get_unconstrained_sizedtypes`
+provide name, type, and shape information.
+The reason that the model reports on both the constrained and unconstrained parameters is the Stan parameter types
+simplex, correlation, covariance, and Cholesky factors have different shapes on the constrained, unconstrained scale.
+For example, a simplex parameter is a vector whose elements sum to one on the constrained scale which means that it has $N-1$ free parameters
```

Comment:
You might want to say $N$ vector (though I don't think math renders here) to give some context as to the N - 1 used later.